rgw: async refcount operate in copy_obj

liangmingyuanneo opened the pull request:
When copy-object objects are called between buckets, the head object is copied in full, other objects only use refcount++ operation. Refcount does not require copying data, but it is processed serially in RGW. It is still time-consuming to go through all rados objects' refcount contained in an RGW object.
The refcount operation itself uses an RPC-like method to construct a request on the RGW side and send the request to the OSD side. After processing the request, the OSD process returns the result. Osd has a special thread pool to process requests, meanwhile different fragments may be sent to multiple OSD nodes for processing. Therefore, if the RGW can send multiple requests at the same time, the processing speed will be greatly improved. So I changed the refcount operation in copy_obj to async mode.

https://tracker.ceph.com/issues/57588

Signed-off-by: Mingyuan Liang <liangmingyuan@baidu.com>

cbodley reviewed (commented):
this is really great, thank you! unfortunately i don't think `librados::AioCompletion` is the best primitive for this - would you be willing to try this with rgw's AioThrottles instead?

anthonyeleven commented on src/common/options/rgw.yaml.in:
```diff
@@ -1913,6 +1913,14 @@ options:
+- name: rgw_max_copy_obj_concurrent_io
+  type: int
+  level: advanced
+  desc: the async refcount io number processed meanwhile in copy_obj
```
Hello, thanks for contributing.  I would like to see this description be more clear.

@cbodley Does the below make sense?

"desc: Number of refcount operations to process concurrently when executing copy_obj"

liangmingyuanneo replied:
ok, this is much better.

liangmingyuanneo commented on src/rgw/rgw_common.h:
```diff
@@ -2551,3 +2551,96 @@ rgw_global_init(const std::map<std::string,std::string> *defaults,
+{
+  assert(!ios.empty());
+  IO &io = ios.front();
+  io.c->wait_for_complete();
```
Year, I will learn it and revise.

cbodley replied:
thank you! please let me know if you have any more questions

liangmingyuanneo commented on src/rgw/rgw_rados.cc:
```diff
@@ -4495,12 +4496,17 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
+    }
 
-      ref_objs.push_back(loc);
+    ret = ref_io_manager.drain_ios();
```
Thank you really for such detailed illustrating, that's very helpful for me. At this please, I think it may does not have to add completed to all_results, because all_results is used for cleanup. What's your opinion?

cbodley replied:
this drains the rest of the requests sent by the refcount loop above. if `drain()` returns any errors, we'll need to run that rollback logic below - and that rollback logic should undo any of the successful writes that we drained here too

so yes, i do think we need to add these completions to `all_results` here

liangmingyuanneo replied:
Yeal, my bad. I revise and commit again, please review when you have time.

liangmingyuanneo commented on src/rgw/rgw_rados.cc:
```diff
@@ -4495,12 +4496,17 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
       ioctx.locator_set_key(loc.loc);
 
-      ret = rgw_rados_operate(dpp, ioctx, loc.oid, &op, null_yield);
+      ret = ref_io_manager.schedule_io(&ioctx, loc.oid, &op, loc);
```
Should we call aio->drain() again before the loop, so we can cleanup obj refcount written, but goto done_ret before calling first drain()?

anthonyeleven commented:
Docs Lgtm

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -4541,21 +4557,33 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
+      static constexpr uint64_t cost = 1; // 1 throttle unit per request
+      static constexpr uint64_t id = 0; // ids unused
+      rgw::AioResultList completed = aio->get(r.obj, rgw::Aio::librados_op(std::move(op), y), cost, id);
+      ret = rgw::check_for_errors(completed);
+      if (ret < 0) {
```
since this is the `done_ret:` error path, we need to make sure we return the original error code - so we can't overwrite `ret` this way

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -4490,17 +4492,31 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
-      auto& ioctx = ref.pool.ioctx();
-      ioctx.locator_set_key(loc.loc);
+      auto obj = svc.rados->obj(miter.get_location().get_raw_obj(store));
+      ret = obj.open(dpp);
```
now that we're using `obj.open()`, we should be able to get rid of the `rgw_rados_ref` stuff above:
```diff
-  rgw_rados_ref ref;
-  ret = get_raw_obj_ref(dpp, miter.get_location().get_raw_obj(store), &ref);
-  if (ret < 0) {
-    return ret;
-  }
```

liangmingyuanneo replied:
year.

cbodley reviewed (commented):
this looks great, how does it work in testing? can we find a way to inject an error to test the rollback logic?

liangmingyuanneo commented:
> this looks great, how does it work in testing? can we find a way to inject an error to test the rollback logic?
I added a unittest, then implemented the refcount++ and rollback logic with BlockingAioThrottle. Please review again at your time.

liangmingyuanneo commented:
> this looks great, how does it work in testing? can we find a way to inject an error to test the rollback logic?

I added a unittest, then implemented the refcount++ and rollback logic with BlockingAioThrottle. Please review again at your time.

liangmingyuanneo commented:
@cbodley

cbodley commented:
> I added a unittest, then implemented the refcount++ and rollback logic with BlockingAioThrottle. Please review again at your time.

nice job getting the test to work! ultimately though, i don't think that makes for a good copyobj regression test because it duplicates the copy logic instead of testing `RGWRados::copy_obj()` directly. we already have https://github.com/ceph/s3-tests to test the s3 copy APIs, and src/test/rgw/test_rgw_throttle.cc for low-level throttle testing

the s3tests for CopyObj won't be able to cover the rollback logic, so i was just looking for confirmation that it works. i added an error after the `drain()` to trigger this rollback:
```diff
@@ -4508,6 +4510,7 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
     rgw::AioResultList completed = aio->drain();
     ret = rgw::check_for_errors(completed);
     all_results.splice(all_results.end(), completed);
+    ret = -EIO; // inject EIO on drain
     if (ret < 0) {
       ldpp_dout(dpp, 0) << "ERROR: failed to drain ios, the error code = " << ret <<dendl;
       goto done_ret;
```
in https://github.com/ceph/s3-tests/pull/473 i added a test case that copies a 16mb object so we could see this ref-counting logic in action. and running that with -EIO error injection, i do see the rollback logic correctly calling `call refcount.put` on each of the 3 tail objects:
```
2022-09-30T12:22:13.945-0400 7f8cd0754640  0 req 5021485027944259865 0.009000298s s3:copy_obj ERROR: failed to drain ios, the error code = -5
2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4443 6.0 6:ffb5a3a3:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_1:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea82a800 con 0x558ce2baf000
2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4444 6.0 6:74a0c32b:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_2:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea960000 con 0x558ce2baf000
2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4445 6.0 6:13df50b3:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_3:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea960400 con 0x558ce2baf000
2022-09-30T12:22:13.946-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4519 ==== osd_op_reply(4443 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_1 [call] v22'62 uv62 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
2022-09-30T12:22:13.947-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4520 ==== osd_op_reply(4444 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_2 [call] v22'63 uv63 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
2022-09-30T12:22:13.947-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4521 ==== osd_op_reply(4445 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_3 [call] v22'64 uv64 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
2022-09-30T12:22:13.947-0400 7f8cd8764640  2 req 5021485027944259865 0.011000365s s3:copy_obj completing
2022-09-30T12:22:13.947-0400 7f8cd8764640  0 WARNING: set_req_state_err err_no=5 resorting to 500
```
could you please remove the test case?

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -4478,6 +4472,8 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
   }
 
+  auto aio = rgw::make_throttle(cct->_conf->rgw_max_copy_obj_concurrent_io, y);
+  rgw::AioResultList all_results;
```
it looks like we could avoid allocating `aio` for the `copy_itself` case?
```c++
  std::unique_ptr<rgw::Aio> aio;
  rgw::AioResultList all_results;
  if (!copy_itself) {
    aio = rgw::make_throttle(cct->_conf->rgw_max_copy_obj_concurrent_io, y);
```

liangmingyuanneo replied:
Well.

liangmingyuanneo commented:
> > I added a unittest, then implemented the refcount++ and rollback logic with BlockingAioThrottle. Please review again at your time.
> 
> nice job getting the test to work! ultimately though, i don't think that makes for a good copyobj regression test because it duplicates the copy logic instead of testing `RGWRados::copy_obj()` directly. we already have https://github.com/ceph/s3-tests to test the s3 copy APIs, and src/test/rgw/test_rgw_throttle.cc for low-level throttle testing
> 
> the s3tests for CopyObj won't be able to cover the rollback logic, so i was just looking for confirmation that it works. i added an error after the `drain()` to trigger this rollback:
> 
> ```diff
> @@ -4508,6 +4510,7 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
>      rgw::AioResultList completed = aio->drain();
>      ret = rgw::check_for_errors(completed);
>      all_results.splice(all_results.end(), completed);
> +    ret = -EIO; // inject EIO on drain
>      if (ret < 0) {
>        ldpp_dout(dpp, 0) << "ERROR: failed to drain ios, the error code = " << ret <<dendl;
>        goto done_ret;
> ```
> 
> in [ceph/s3-tests#473](https://github.com/ceph/s3-tests/pull/473) i added a test case that copies a 16mb object so we could see this ref-counting logic in action. and running that with -EIO error injection, i do see the rollback logic correctly calling `call refcount.put` on each of the 3 tail objects:
> 
> ```
> 2022-09-30T12:22:13.945-0400 7f8cd0754640  0 req 5021485027944259865 0.009000298s s3:copy_obj ERROR: failed to drain ios, the error code = -5
> 2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4443 6.0 6:ffb5a3a3:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_1:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea82a800 con 0x558ce2baf000
> 2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4444 6.0 6:74a0c32b:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_2:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea960000 con 0x558ce2baf000
> 2022-09-30T12:22:13.945-0400 7f8cd0754640  1 -- 192.168.245.128:0/352496449 --> [v2:192.168.245.128:6800/850826,v1:192.168.245.128:6801/850826] -- osd_op(unknown.0.0:4445 6.0 6:13df50b3:::eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_3:head [call refcount.put in=84b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e22) v8 -- 0x558cea960400 con 0x558ce2baf000
> 2022-09-30T12:22:13.946-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4519 ==== osd_op_reply(4443 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_1 [call] v22'62 uv62 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
> 2022-09-30T12:22:13.947-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4520 ==== osd_op_reply(4444 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_2 [call] v22'63 uv63 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
> 2022-09-30T12:22:13.947-0400 7f8d9c1c6640  1 -- 192.168.245.128:0/352496449 <== osd.0 v2:192.168.245.128:6800/850826 4521 ==== osd_op_reply(4445 eb6b7daa-f6c3-46a8-9a92-5e649c93b954.4137.7__shadow_.T2kNhevtg8qKE7paX1WOXBXh1kzQRkM_3 [call] v22'64 uv64 ondisk = 0) v8 ==== 230+0+0 (crc 0 0 0) 0x558ce7a14d80 con 0x558ce2baf000
> 2022-09-30T12:22:13.947-0400 7f8cd8764640  2 req 5021485027944259865 0.011000365s s3:copy_obj completing
> 2022-09-30T12:22:13.947-0400 7f8cd8764640  0 WARNING: set_req_state_err err_no=5 resorting to 500
> ```
> 
> could you please remove the test case?

ok.

cbodley reviewed (approved):
thanks! passed qa in https://pulpito.ceph.com/cbodley-2022-09-30_16:47:21-rgw-wip-cbodley-testing-distro-default-smithi/

cbodley commented:
jenkins test api

cbodley commented:
@liangmingyuanneo do you need this fix on the pacific or quincy release? we generally don't backport this kind of performance improvement

liangmingyuanneo commented:
> @liangmingyuanneo do you need this fix on the pacific or quincy release? we generally don't backport this kind of performance improvement

No. So I don't backport it either now.
